scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+	taskc->slice_ns = slice;
+	return slice;
+}
+
+static bool transit_task_stat(struct task_ctx *taskc, int tgt_stat)
+{
+	/*
+	 * Update task loads only when the state transition is valid.
+	 *
+	 * initial state
+	 * -------------
+	 *     |
+	 *    \/
+	 * [STOPPING] --> [ENQ] --> [RUNNING]
+	 *    /\                        |
+	 *    |                         |
+	 *    +-------------------------+
+	 */
+	const static int valid_tgt_stat[] = {
+		LAVD_TASK_STAT_ENQ,
+		LAVD_TASK_STAT_RUNNING,
+		LAVD_TASK_STAT_STOPPING,
+	};
+	int src_stat = taskc->stat;
+
+	if (src_stat < _LAVD_TASK_STAT_MIN || src_stat > _LAVD_TASK_STAT_MAX) {
+		scx_bpf_error("Invalid task state: %d", src_stat);
+		return false;
+	}
+
+	if (valid_tgt_stat[src_stat] == tgt_stat) {
+		taskc->stat = tgt_stat;
+		return true;
+	}
+
+	return false;
+}
+
+static void update_stat_for_enq(struct task_struct *p, struct task_ctx *taskc,
+				struct cpu_ctx *cpuc)
+{
+	/*
+	 * Reflect task's load immediately.
+	 */
+	taskc->load_actual = calc_task_load_actual(taskc);
+	cpuc->load_actual += taskc->load_actual;
+	cpuc->load_ideal  += get_task_load_ideal(p);
+}
+
+static void update_stat_for_run(struct task_struct *p, struct task_ctx *taskc,
+				struct cpu_ctx *cpuc)
+{
+	u64 now, wait_period, interval;
+
+	/*
+	 * Since this is the start of a new schedule for @p, we update run
+	 * frequency in a second using an exponential weighted moving average.
+	 */
+	now = bpf_ktime_get_ns();
+	wait_period = now - taskc->last_stop_clk;
+	interval = taskc->run_time_ns + wait_period;
+	taskc->run_freq = calc_avg_freq(taskc->run_freq, interval);
+
+	/*
+	 * Update task state when starts running.
+	 */
+	taskc->last_start_clk = now;
+}
+
+static void update_stat_for_stop(struct task_struct *p, struct task_ctx *taskc,
+				 struct cpu_ctx *cpuc)
+{
+	u64 now, run_time_ns;
+
+	/*
+	 * If the task was runnable before governed by this scheduler, ignore
+	 * this run.
+	 */
+	now = bpf_ktime_get_ns();
+	if (now < taskc->last_start_clk)
+		return;
+
+	/*
+	 * When stopped, reduce the per-CPU task load. Per-CPU task load will
+	 * be aggregated periodically at update_sys_cpu_load().
+	 */
+	cpuc->load_actual -= taskc->load_actual;
+	cpuc->load_ideal  -= get_task_load_ideal(p);
+
+	/*
+	 * Update task's run_time.
+	 */
+	run_time_ns = now - taskc->last_start_clk;
+	taskc->run_time_ns = calc_avg(taskc->run_time_ns, run_time_ns);
+	taskc->last_stop_clk = now;
+}
+
+static void calc_when_to_run(struct task_struct *p, struct task_ctx *taskc,
+			     u64 enq_flags)
+{
+	/*
+	 * Before enqueueing a task to a run queue, we should decide when a
+	 * task should be scheduled. It is determined by two factors: how
+	 * urgent it is - vdeadline_delta_ns - and when it becomes eligible if
+	 * overscheduled - eligible_time_ns.
+	 */
+	calc_virtual_dealine_delta(p, taskc, enq_flags);
+	calc_eligible_delta(p, taskc);
+}
+
+static bool put_local_rq(struct task_struct *p, struct task_ctx *taskc,
+			 s32 cpu_id, u64 enq_flags)
```

Comment:
Do we really need to pass cpu_id in `put_local_rq()`? Being a local dispatch it shouldn't require a target cpu right?

Reply:
That's a good catch. Hehe, that's a side effect of multiple iterations. I will fix it.